serial-multiplexer/src/main.rs

```diff
@@ -42,130 +54,139 @@ fn main() {
+    receiver_processors: Vec<SerialConnectionReceiverProcessor>,
+    senders: Vec<SerialConnectionSender>,
+    main_bus_receiver: Receiver<DataBlock>,
     multiplexed_port: &mut TTYPort,
-    serial_ports: &mut HashMap<u32, SerialEntry>,
 ) {
-    let mut multiplexed_port_clone = multiplexed_port.try_clone_native().unwrap();
+    sender_processors.into_iter().for_each(|f| {
+        std::thread::spawn(move || {
+            f.process_loop();
+        });
+    });
 
-    let mut serial_ports_clone: HashMap<u32, SerialEntry> = HashMap::new();
-
-    serial_ports.iter().for_each(|port| {
-        serial_ports_clone.insert(
-            port.0.clone(),
-            SerialEntry { 
-                name: port.1.name.clone(), 
-                device: port.1.device.try_clone_native().unwrap(), 
-                id: port.1.id
-            },
-        );
+    receiver_processors.into_iter().for_each(|f| {
+        std::thread::spawn(move || {
+            f.process_loop();
+        });
     });
 
-    std::thread::spawn(move || {
-        let mut local_map = serial_ports_clone;
-        
+    let mut multiplexed_port_clone = multiplexed_port.try_clone_native().unwrap();
 
+    std::thread::spawn(move || {
         loop {
-            let mut any_data = false;
-
-            local_map.iter_mut().for_each(|port| {
-                let mut buff = [0u8; 255];
-                if let Ok(bytes_to_read) = port.1.device.bytes_to_read()
-                {
-                    if bytes_to_read > 0
-                    {
-                        if let Ok(bytes_read) = port.1.device.read(&mut buff) {
-                            if bytes_read > 0 {
-                                let mut mini_buff = [0u8; 2];
-                                mini_buff[0] = port.0.clone() as u8;
-                                mini_buff[1] = bytes_read as u8;
-
-                                multiplexed_port_clone.write_all(&mini_buff).unwrap();
-                                multiplexed_port_clone.write_all(&buff[..bytes_read]).unwrap();
-                                #[cfg(debug_assertions)]
-                                {
-                                    println!("Sent {} bytes for device {}", bytes_read, port.0);
-                                }
-                            }
-                        }
-
-                        any_data = true;
-                    }
-                }
-            });
+            let data = main_bus_receiver.recv().unwrap();
 
-            // TODO: Hacky fix to make this not eat up all the cpu. Use st3 in the future
-            if !any_data
-            {
-                thread::sleep(Duration::from_micros(100));
-            }
+            let mut mini_buff = [0u8; 2];
+            mini_buff[0] = data.id as u8;
+            mini_buff[1] = data.data.len() as u8;
+
+            multiplexed_port_clone.write_all(&mini_buff).unwrap();
+            multiplexed_port_clone.write_all(&data.data).unwrap();
```

Comment:
Maybe we can gracefully handle most other `.expect()` and `.unwrap()`s?

Reply:
Done, also added retries on opening the multiplexed port.

This entire thing does collapse whenever there's an IO error instead of a device unavailable error, but that's a problem for later